pycloudlib/ec2/instance.py

```diff
@@ -81,9 +105,67 @@ def add_network_interface(self) -> str:
         See the AWS documentation for more info:
         https://boto3.readthedocs.io/en/latest/reference/services/ec2.html?#EC2.Client.create_network_interface
         https://boto3.readthedocs.io/en/latest/reference/services/ec2.html?#EC2.Client.attach_network_interface
+
+        Args:
+            ipv4_address_count: number of private ipv4s
+            ipv6_address_count: number of private ipv6s
+            ipv4_public_ip_count: number of public ips associated to private
+                ips.
+            associate_public_ip: created boto3 instance object
+        Returns: private primary ip
         """
         self._log.debug("adding network interface to %s", self.id)
-        interface_id = self._create_network_interface()
+        interface_id = self._create_network_interface(ipv6_address_count)
+        if ipv4_public_ip_count is not None:
+            if (
+                ipv4_address_count is None
+                or ipv4_address_count < ipv4_public_ip_count
+            ):
+                raise PycloudlibError(
+                    f"Invalid configuration:"
+                    f" `ipv4_public_ip_count={ipv4_public_ip_count}` cannot be"
+                    f" greater than `ipv4_address_count={ipv4_address_count}`"
+                )
+
+        if ipv4_address_count is not None and ipv4_address_count > 1:
```

Comment:
if ipv4_address_count == 1, then we aren't going to call allocate_address/associate_address below. In checking the launched instances,I'm already seeing two private ips before any additional allocation based on ipv4_address_count > 1. Do we already expect that initial IPv4 private address to be allocated and associated already without needing to call allocate_address or associate_address within the for i in range(ipv4_address_count) below? ? 

Also, if ipv4_address_count  == 2, we'll call allocate_address/associate_address twice because our range below is for i in range(ipv4_address_count). Does this actually add 2 extra private IPs to this interface even though it may already have 1 private IP associated with it? Should that range actually be `range(ipv4_address_count -1)` instead?

Also, in testing this PR with the corresponding draft PR in cloud-init/ec2-ipv6 on the integration test  tests/integration_tests/test_networking.py::test_ec2_multi_nic_reboot. I'm getting botocore errors that we are allocating too many addresses. I'm wondering if that's just on my account, or due to the approach we are taking to continue to allocate more addresses than are typically allowed for an instance 
```
E           botocore.exceptions.ClientError: An error occurred (AddressLimitExceeded) when calling the AllocateAddress operation: The maximum number of addresses has been reached.
```

Reply:
I messed up, by copy-pasting, `ipv4_address_count` and `ipv4_public_count`. Expectations are that the default NIC (the one associated with the instance at launch) will have a private and public ipv4 by default so pycloudlib does not have to create them. I have fixed it, added comments and improved the documentation to reflect that.